File: app/src/main/java/com/ironmeta/one/ui/regionselector2/ServerListActivity.java
package com.ironmeta.one.ui.regionselector2;

import android.app.Activity;
import android.os.Bundle;
import androidx.appcompat.widget.Toolbar;

import androidx.annotation.Nullable;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.DefaultItemAnimator;
import androidx.recyclerview.widget.DividerItemDecoration;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.ironmeta.one.R;
import com.ironmeta.one.coreservice.CoreServiceManager;
import com.ironmeta.one.report.AppReport;
import com.ironmeta.one.report.ReportConstants;
import com.ironmeta.one.ui.common.CommonAppCompatActivity;
import com.ironmeta.one.ui.common.CommonDialog;
import com.ironmeta.one.ui.support.LegalManager;

import java.util.concurrent.atomic.AtomicReference;

public class ServerListActivity extends CommonAppCompatActivity {
    private ServerListViewModel mServerListViewModel;
    private final AtomicReference<CommonDialog> mLegalNoticeDialog = new AtomicReference<>();

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_server_list);

        RecyclerView serverListRV = findViewById(R.id.recycler_view);

        mServerListViewModel = new ViewModelProvider(this).get(ServerListViewModel.class);

        ((Toolbar) findViewById(R.id.toolbar)).setNavigationOnClickListener(v -> onBackPressed());

        getWindow().setBackgroundDrawableResource(R.color.white);

        LinearLayoutManager linearLayoutManager = new LinearLayoutManager(this);
        linearLayoutManager.setOrientation(RecyclerView.VERTICAL);
        ServerListRecyclerViewAdapter serverListRecyclerViewAdapter = new ServerListRecyclerViewAdapter(ServerListActivity.this, vpnServerRegion -> {
            AppReport.setConnectionSource(ReportConstants.AppReport.SOURCE_CONNECTION_PAGE_SERVER);
            AppReport.reportToConnection();
            CoreServiceManager.getInstance(getApplicationContext()).connect(vpnServerRegion);
            finish();
        });

        serverListRV.setLayoutManager(linearLayoutManager);
        serverListRV.setAdapter(serverListRecyclerViewAdapter);
        serverListRV.addItemDecoration(new DividerItemDecoration(this, DividerItemDecoration.VERTICAL));
        serverListRV.setItemAnimator(new DefaultItemAnimator());
        mServerListViewModel.getVPNServerRegionList().observe(this, serverListRecyclerViewAdapter::setVPNServerRegions);

        LegalManager.getInstance(getApplicationContext()).getInLegalRegionAsLiveData().observe(this, this::legalRegionObserverHandler);

        initServersRefresh();
    }

    private void initServersRefresh() {
        mServerListViewModel.getServersRefreshingAsLiveData().observe(this, serversRefreshingResult -> {
            if (serversRefreshingResult != null && serversRefreshingResult) {
                showLoading(getResources().getString(R.string.vs_feature_region_refresh_tips_loading), false);
                return;
            }
            cancelLoading();
        });
        findViewById(R.id.servers_refresh_btn).setOnClickListener(v -> mServerListViewModel.refreshServers());
    }

    private void legalRegionObserverHandler(boolean isLegal) {
        if (isLegal) return;

        final CommonDialog dialog = new CommonDialog(this);
        if (!mLegalNoticeDialog.compareAndSet(null, dialog)) return;
        dialog.setCancelable(false);
        dialog.setOnlyOKButton(true);
        dialog.setTitle(getString(R.string.vs_legal_notices_dialog_title));
        dialog.setMessage(getString(R.string.vs_legal_notices_dialog_content));
        dialog.setOKButton(getString(R.string.vs_common_dialog_ok_button));
        dialog.setOkOnclickListener(() -> {
            mLegalNoticeDialog.set(null);
            System.exit(0);
        });
        dialog.show();
    }

    @Override
    public void onBackPressed() {
        setResult(Activity.RESULT_OK);
        super.onBackPressed();
    }
}

File: app/src/main/java/com/ironmeta/one/ui/regionselector2/ServerListActivity.kt
package com.ironmeta.one.ui.regionselector2

import android.os.Bundle
import android.view.View
import androidx.appcompat.widget.Toolbar
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.lifecycleScope
import androidx.recyclerview.widget.DefaultItemAnimator
import androidx.recyclerview.widget.DividerItemDecoration
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.ironmeta.one.R
import com.ironmeta.one.coreservice.CoreServiceManager.Companion.getInstance
import com.ironmeta.one.report.RequestVpnPermissionContract
import com.ironmeta.one.report.VpnReporter
import com.ironmeta.one.ui.common.CommonAppCompatActivity
import com.ironmeta.one.ui.common.CommonDialog
import com.ironmeta.one.ui.support.LegalManager
import com.sdk.ssmod.api.http.beans.FetchResponse
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import java.util.concurrent.atomic.AtomicReference
import kotlin.coroutines.resume
import kotlin.coroutines.suspendCoroutine
import kotlin.jvm.internal.Ref

class ServerListActivity : CommonAppCompatActivity() {
    private var mServerListViewModel: ServerListViewModel? = null
    private val mLegalNoticeDialog = AtomicReference<CommonDialog?>()
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_server_list)
        val serverListRV = findViewById<RecyclerView>(R.id.recycler_view)
        mServerListViewModel = ViewModelProvider(this).get(ServerListViewModel::class.java)
        (findViewById<View>(R.id.toolbar) as Toolbar).setNavigationOnClickListener { v: View? -> onBackPressed() }
        window.setBackgroundDrawableResource(R.color.white)
        val linearLayoutManager = LinearLayoutManager(this)
        linearLayoutManager.orientation = RecyclerView.VERTICAL
        val serverListRecyclerViewAdapter =
            ServerListRecyclerViewAdapter(this@ServerListActivity) { vpnServerRegion: FetchResponse.ServerZone? ->
                VpnReporter.reportToStartConnect(VpnReporter.PARAM_VALUE_FROM_SERVER_LIST)
                lifecycleScope.launch {
                    if (!requestVpnPermission(this@ServerListActivity)) {
                        return@launch
                    }
                    VpnReporter.reportStartConnect(VpnReporter.PARAM_VALUE_FROM_SERVER_LIST)
                    getInstance(applicationContext).connect(vpnServerRegion)
                    launch(Dispatchers.Main) { finish() }
                }
            }
        serverListRV.layoutManager = linearLayoutManager
        serverListRV.adapter = serverListRecyclerViewAdapter
        serverListRV.addItemDecoration(DividerItemDecoration(this, DividerItemDecoration.VERTICAL))
        serverListRV.itemAnimator = DefaultItemAnimator()
        mServerListViewModel!!.vPNServerRegionList.observe(this) { vpnServerRegions: FetchResponse? ->
            serverListRecyclerViewAdapter.setVPNServerRegions(
                vpnServerRegions
            )
        }
        LegalManager.getInstance(applicationContext).inLegalRegionAsLiveData.observe(this) { isLegal: Boolean ->
            legalRegionObserverHandler(
                isLegal
            )
        }
        initServersRefresh()
    }

    private var requestVpnPermissionCallbackRef = Ref.ObjectRef<(Boolean) -> Unit>()
    private val requestVpnPermissionContract = registerForActivityResult(RequestVpnPermissionContract()) {
            requestVpnPermissionCallbackRef.element?.invoke(it)
        }
    private suspend fun requestVpnPermission(activity: CommonAppCompatActivity) = suspendCoroutine<Boolean> { continuation ->
        requestVpnPermissionCallbackRef.element = {
            requestVpnPermissionCallbackRef.element = null
            continuation.resume(it)
        }
        requestVpnPermissionContract.launch(Unit)
    }

    private fun initServersRefresh() {
        mServerListViewModel?.serversRefreshingAsLiveData?.observe(this) { serversRefreshingResult: Boolean? ->
                if (serversRefreshingResult != null && serversRefreshingResult) {
                    showLoading(resources.getString(R.string.vs_feature_region_refresh_tips_loading), false)
                    return@observe
                }
                cancelLoading()
            }
        findViewById<View>(R.id.servers_refresh_btn).setOnClickListener { v: View? -> mServerListViewModel!!.refreshServers() }
    }

    private fun legalRegionObserverHandler(isLegal: Boolean) {
        if (isLegal) return
        val dialog = CommonDialog(this)
        if (!mLegalNoticeDialog.compareAndSet(null, dialog)) return
        dialog.setCancelable(false)
        dialog.setOnlyOKButton(true)
        dialog.setTitle(getString(R.string.vs_legal_notices_dialog_title))
        dialog.setMessage(getString(R.string.vs_legal_notices_dialog_content))
        dialog.setOKButton(getString(R.string.vs_common_dialog_ok_button))
        dialog.setOkOnclickListener {
            mLegalNoticeDialog.set(null)
            System.exit(0)
        }
        dialog.show()
    }

    override fun onBackPressed() {
        setResult(RESULT_OK)
        super.onBackPressed()
    }
}